doc/design/refactor_discuess/net_design.md

```diff
@@ -0,0 +1,65 @@
+# Network Design
+
+`Network` is the container and controller of a set of operators in a network, users can use `Network.addOp` to add operators into a network, 
+and use `Network.runOps` to run all the operators in the network.
+
+The `Network` will
+
+- manage all the operators contained in the network.
+- not own any `Variable`.
+
+# API
+
+To make the `Network` extendibe, a base class is defined like this
+
+```c++
+// The minimum a network should be implemented.
+class BaseNetwork {
+public:
+  BaseNetwork(const NetDef& def, Scope *scope);
+
+  // run all the operators and return success(true) or not.
+  virtual bool Run() = 0;
+
+protected:
+  // the input variables feed into the network.
+  std::vector<Variable*> external_inputs_;
```

Comment:
It seems that a net could read and/or write a variable, so we might have *read-only* and *writable* variables, other than *input* and *output* variables, associated with it?

I am not sure if we should distinguish them, or just have `vars_`.